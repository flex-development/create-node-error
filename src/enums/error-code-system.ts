/**
 * @file Enums - SystemErrorCode
 * @module errnode/enums/SystemErrorCode
 */

/**
 * System error codes.
 *
 * @see https://github.com/nodejs/node/blob/v19.3.0/deps/uv/include/uv.h#L72-L154
 *
 * @enum {Uppercase<string>}
 */
enum SystemErrorCode {
  E2BIG = 'E2BIG',
  EACCES = 'EACCES',
  EADDRINUSE = 'EADDRINUSE',
  EADDRNOTAVAIL = 'EADDRNOTAVAIL',
  EAFNOSUPPORT = 'EAFNOSUPPORT',
  EAGAIN = 'EAGAIN',
  EAI_ADDRFAMILY = 'EAI_ADDRFAMILY',
  EAI_AGAIN = 'EAI_AGAIN',
  EAI_BADFLAGS = 'EAI_BADFLAGS',
  EAI_BADHINTS = 'EAI_BADHINTS',
  EAI_CANCELED = 'EAI_CANCELED',
  EAI_FAIL = 'EAI_FAIL',
  EAI_FAMILY = 'EAI_FAMILY',
  EAI_MEMORY = 'EAI_MEMORY',
  EAI_NODATA = 'EAI_NODATA',
  EAI_NONAME = 'EAI_NONAME',
  EAI_OVERFLOW = 'EAI_OVERFLOW',
  EAI_PROTOCOL = 'EAI_PROTOCOL',
  EAI_SERVICE = 'EAI_SERVICE',
  EAI_SOCKTYPE = 'EAI_SOCKTYPE',
  EALREADY = 'EALREADY',
  EBADF = 'EBADF',
  EBADMSG = 'EBADMSG',
  EBUSY = 'EBUSY',
  ECANCELED = 'ECANCELED',
  ECHARSET = 'ECHARSET',
  ECHILD = 'ECHILD',
  ECONNABORTED = 'ECONNABORTED',
  ECONNREFUSED = 'ECONNREFUSED',
  ECONNRESET = 'ECONNRESET',
  EDEADLK = 'EDEADLK',
  EDESTADDRREQ = 'EDESTADDRREQ',
  EDOM = 'EDOM',
  EDQUOT = 'EDQUOT',
  EEXIST = 'EEXIST',
  EFAULT = 'EFAULT',
  EFBIG = 'EFBIG',
  EFTYPE = 'EFTYPE',
  EHOSTDOWN = 'EHOSTDOWN',
  EHOSTUNREACH = 'EHOSTUNREACH',
  EIDRM = 'EIDRM',
  EILSEQ = 'EILSEQ',
  EINPROGRESS = 'EINPROGRESS',
  EINTR = 'EINTR',
  EINVAL = 'EINVAL',
  EIO = 'EIO',
  EISCONN = 'EISCONN',
  EISDIR = 'EISDIR',
  ELOOP = 'ELOOP',
  EMFILE = 'EMFILE',
  EMLINK = 'EMLINK',
  EMSGSIZE = 'EMSGSIZE',
  EMULTIHOP = 'EMULTIHOP',
  ENAMETOOLONG = 'ENAMETOOLONG',
  ENETDOWN = 'ENETDOWN',
  ENETRESET = 'ENETRESET',
  ENETUNREACH = 'ENETUNREACH',
  ENFILE = 'ENFILE',
  ENOBUFS = 'ENOBUFS',
  ENODATA = 'ENODATA',
  ENODEV = 'ENODEV',
  ENOENT = 'ENOENT',
  ENOEXEC = 'ENOEXEC',
  ENOLCK = 'ENOLCK',
  ENOLINK = 'ENOLINK',
  ENOMEM = 'ENOMEM',
  ENOMSG = 'ENOMSG',
  ENONET = 'ENONET',
  ENOPROTOOPT = 'ENOPROTOOPT',
  ENOSPC = 'ENOSPC',
  ENOSR = 'ENOSR',
  ENOSTR = 'ENOSTR',
  ENOSYS = 'ENOSYS',
  ENOTCONN = 'ENOTCONN',
  ENOTDIR = 'ENOTDIR',
  ENOTEMPTY = 'ENOTEMPTY',
  ENOTSOCK = 'ENOTSOCK',
  ENOTSUP = 'ENOTSUP',
  ENOTTY = 'ENOTTY',
  ENXIO = 'ENXIO',
  EOF = 'EOF',
  EOPNOTSUPP = 'EOPNOTSUPP',
  EOVERFLOW = 'EOVERFLOW',
  EPERM = 'EPERM',
  EPIPE = 'EPIPE',
  EPROTO = 'EPROTO',
  EPROTONOSUPPORT = 'EPROTONOSUPPORT',
  EPROTOTYPE = 'EPROTOTYPE',
  ERANGE = 'ERANGE',
  EREMOTEIO = 'EREMOTEIO',
  EROFS = 'EROFS',
  ESHUTDOWN = 'ESHUTDOWN',
  ESOCKTNOSUPPORT = 'ESOCKTNOSUPPORT',
  ESPIPE = 'ESPIPE',
  ESRCH = 'ESRCH',
  ESTALE = 'ESTALE',
  ETIME = 'ETIME',
  ETIMEDOUT = 'ETIMEDOUT',
  ETXTBSY = 'ETXTBSY',
  EUNATCH = 'EUNATCH',
  EWOULDBLOCK = 'EWOULDBLOCK',
  EXDEV = 'EXDEV',
  UNKNOWN = 'UNKNOWN'
}

export default SystemErrorCode
