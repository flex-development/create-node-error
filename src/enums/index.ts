/**
 * @file Entry Point - Enums
 * @module errnode/enums
 */

export { default as codes } from './codes'
export { default as syscodes } from './syscodes'
