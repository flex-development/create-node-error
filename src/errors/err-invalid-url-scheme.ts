/**
 * @file Errors - ERR_INVALID_URL_SCHEME
 * @module errnode/errors/ERR_INVALID_URL_SCHEME
 * @see https://github.com/nodejs/node/blob/v22.8.0/lib/internal/errors.js#L1543-L1552
 */

import E from '#e'
import { codes } from '#src/enums'
import type { NodeError, NodeErrorConstructor } from '#src/interfaces'
import formatList from '#src/utils/format-list'
import { ok } from 'devlop'

/**
 * `ERR_INVALID_URL_SCHEME` schema.
 *
 * @see {@linkcode NodeError}
 * @see https://nodejs.org/api/errors.html#err_invalid_url_scheme
 *
 * @extends {NodeError<codes.ERR_INVALID_URL_SCHEME>}
 * @extends {TypeError}
 */
interface ErrInvalidUrlScheme
  extends NodeError<codes.ERR_INVALID_URL_SCHEME>, TypeError {}

/**
 * `ERR_INVALID_URL_SCHEME` message arguments.
 */
type Args = [expected: string | readonly string[]]

/**
 * `ERR_INVALID_URL_SCHEME` constructor.
 *
 * @see {@linkcode Args}
 * @see {@linkcode ErrInvalidUrlScheme}
 * @see {@linkcode NodeErrorConstructor}
 *
 * @extends {NodeErrorConstructor<ErrInvalidUrlScheme,Args>}
 */
interface ErrInvalidUrlSchemeConstructor
  extends NodeErrorConstructor<ErrInvalidUrlScheme, Args> {
  /**
   * Create a new `ERR_INVALID_URL_SCHEME` error.
   *
   * @see {@linkcode ErrInvalidUrlScheme}
   *
   * @param {ReadonlyArray<string> | string} expected
   *  Expected scheme or list of expected schemes
   * @return {ErrInvalidUrlScheme}
   */
  new (expected: string | readonly string[]): ErrInvalidUrlScheme
}

/**
 * `ERR_INVALID_URL_SCHEME` model.
 *
 * Thrown when an attempt is made to use a URL of an incompatible scheme
 * (protocol).
 *
 * @see {@linkcode ErrInvalidUrlSchemeConstructor}
 *
 * @type {ErrInvalidUrlSchemeConstructor}
 *
 * @class
 */
const ERR_INVALID_URL_SCHEME: ErrInvalidUrlSchemeConstructor = E(
  codes.ERR_INVALID_URL_SCHEME,
  TypeError,
  /**
   * @param {ReadonlyArray<string> | string} expected
   *  Expected scheme or list of expected schemes
   * @return {string}
   *  Error message
   */
  function message(expected: string | readonly string[]): string {
    ok(expected.length, 'expected `expected` to not be empty')

    /**
     * Error message.
     *
     * @var {string} message
     */
    let message: string = 'The URL must be '

    message += Array.isArray(expected) && expected.length > 1
      ? 'one of scheme'
      : 'of scheme'

    message += ` ${formatList(expected, 'or')}`

    return message
  }
)

export {
  ERR_INVALID_URL_SCHEME as default,
  type ErrInvalidUrlScheme,
  type Args as ErrInvalidUrlSchemeArgs,
  type ErrInvalidUrlSchemeConstructor
}
