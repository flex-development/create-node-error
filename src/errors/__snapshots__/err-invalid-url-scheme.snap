// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:errors/ERR_INVALID_URL_SCHEME > should build message from parameters (0) 1`] = `"The URL must be of scheme file"`;

exports[`unit:errors/ERR_INVALID_URL_SCHEME > should build message from parameters (1) 1`] = `"The URL must be one of scheme file or node"`;
